### PR TITLE
[CELEBORN-502][REFACTOR] Merge GetBlacklistResponse to HeartbeatFromApplication

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Merge GetBlacklistResponse to HeartbeatFromApplication

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@RexXiong @waitinfuture

**Comment by RexXiong** [comment]
Agree with the merge Idea, and shell we consider introducing a workerListener interface to monitor worker status changes, and then we can do something. In [CELEBORN-479](https://github.com/apache/incubator-celeborn/pull/1405) I need register a worker listener to monitor unavailable workers as i cannot pass resourceTracker(belong to Flink plugin) to lifecycleManagers. If that we can have shutdownWorkerListener for CommitManager, UnknownWorkerListener for Flink, workerFailureListener for blackList etc.  What do you think?

**Comment by AngersZhuuuu** [comment]
> Agree with the merge Idea, and shell we consider introducing a workerListener interface to monitor worker status changes, and then we can do something. In [CELEBORN-479](https://github.com/apache/incubator-celeborn/pull/1405) I need register a worker listener to monitor unavailable workers as i cannot pass resourceTracker(belong to Flink plugin) to lifecycleManagers. If that we can have shutdownWorkerListener for CommitManager, UnknownWorkerListener for Flink, workerFailureListener for blackList etc. What do you think?

SGTM， we can add this listener later, it will make the code more clear

**Comment by AngersZhuuuu** [comment]
ping @RexXiong @waitinfuture

**Comment by waitinfuture** [comment]
> ping @RexXiong @waitinfuture

I will review this PR late today

**Comment by waitinfuture** [comment]
I'm afraid this PR will cause back compatibility issues. For example old client invoke ```rssHARetryClient.askSync[GetBlacklistResponse](message, classOf[GetBlacklistResponse])```, but server don't know ```GetBlacklist```.
How about retaining GetBlacklist/GetBlacklistResponse message, and add new HeartBeat/HeartBeatResponse messages?

File: client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  val blacklist = new ShuffleFailedWorkers()
+  private val shuttingWorkers: JSet[WorkerInfo] = new JHashSet[WorkerInfo]()
+
+  def blacklistPartition(
+      shuffleId: Int,
+      oldPartition: PartitionLocation,
+      cause: StatusCode): Unit = {
+    val failedWorker = new ShuffleFailedWorkers()
+
+    def blacklistPartitionWorker(
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala:48]: Done

File: client/src/main/scala/org/apache/celeborn/client/ApplicationHeartbeater.scala [hunk 2]
```diff
@@ -48,9 +51,17 @@ class ApplicationHeartbeater(
...
+                appId,
+                tmpTotalWritten,
+                tmpFileCount,
+                blacklistManager.blacklist.asScala.keys.toList.asJava,
+                ZERO_UUID)
+            val response = requestHeartbeat(appHeartbeat)
+            if (response.statusCode == StatusCode.SUCCESS) {
+              logDebug("Successfully send app heartbeat.")
+            }
+            blacklistManager.handleHeartbeatResponse(response)
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/ApplicationHeartbeater.scala]: Done

File: client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.celeborn.common.internal.Logging
+import org.apache.celeborn.common.meta.WorkerInfo
+import org.apache.celeborn.common.protocol.PartitionLocation
+import org.apache.celeborn.common.protocol.message.ControlMessages.HeartbeatFromApplicationResponse
+import org.apache.celeborn.common.protocol.message.StatusCode
+
+class LifecycleBlacklistManager(
+    conf: CelebornConf,
+    lifecycleManager: LifecycleManager,
+    commitManager: CommitManager) extends Logging {
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala]: > It's strange to add CommitManager in LifecycleBlacklistManager. Seems your previous discussion will make the code cleaner, let's do it later.

How about the current code? And later we will implement a listener in each component, the we don't need to involve each other.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        }
+      }
+    }
+  }
+
+  def handleHeartbeatResponse(res: HeartbeatFromApplicationResponse): Unit = {
+    if (res.statusCode == StatusCode.SUCCESS) {
+      logInfo(s"Received Blacklist from Master, blacklist: ${res.blacklist} " +
+        s"unknown workers: ${res.unknownWorkers}, shutdown workers: ${res.shutdownWorkers}")
+      resolveShutdownWorkers(res.shutdownWorkers)
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleBlacklistManager.scala]: Done

**Comment by AngersZhuuuu** [comment]
@waitinfuture @RexXiong How about current? I am wondering if I should remove `shuttingWorkers` in GetBlacklistResponse...

File: common/src/main/proto/TransportMessages.proto [hunk 5]
```diff
@@ -39,8 +39,8 @@ enum MessageType {
   APPLICATION_LOST = 18;
   APPLICATION_LOST_RESPONSE = 19;
   HEARTBEAT_FROM_APPLICATION = 20;
-  GET_BLACKLIST = 21;
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply common/src/main/proto/TransportMessages.proto]: Done

**Comment by AngersZhuuuu** [comment]
> > @waitinfuture @RexXiong How about current? I am wondering if I should remove `shuttingWorkers` in GetBlacklistResponse...
> 
> I agree to remove it directly, as we do not need to maintain compatibility during major version development.

Done

**Comment by AngersZhuuuu** [comment]
Any more suggestion? ping @waitinfuture @RexXiong

**Review by RexXiong — APPROVED** [review]
LGTM! cc @waitinfuture

**Review by waitinfuture — APPROVED** [review]
LGTM modulo a nit, thanks!

File: client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.celeborn.client
+
+import java.util.{function, HashSet => JHashSet, List => JList, Set => JSet}
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala]: Removed

**Pull request merged**